addons/account/tests/test_product.py

```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+
+from odoo.addons.account.tests.common import AccountTestInvoicingCommon
+from odoo.tests import tagged
+
+
+@tagged('post_install', '-at_install')
+class TestAccountPaymentTerms(AccountTestInvoicingCommon):
+    @classmethod
+    def setUpClass(cls, chart_template_ref=None):
+        super().setUpClass(chart_template_ref=chart_template_ref)
+
+    def test_multi_company_product_tax(self):
+        """ Ensure default taxes are set for all companies on products with no company set. """
+        product_without_company = self.env['product.template'].create({
+            'name': 'Product Without a Company',
+        })
+        product_with_company = self.env['product.template'].create({
+            'name': 'Product With a Company',
+            'company_id': self.company_data['company'].id,
+        })
+        companies = self.env['res.company'].sudo().search([])
+        # Product should have all the default taxes of the other companies.
+        self.assertRecordValues(product_without_company.sudo(), [{
+            'taxes_id': companies.account_sale_tax_id.ids,
+            'supplier_taxes_id': companies.account_purchase_tax_id.ids,
+        }])
+        # Product should have only the default tax of the company it belongs to.
+        self.assertRecordValues(product_with_company, [{
+            'taxes_id': self.company_data['company'].account_sale_tax_id.ids,
```

Comment:
to be fair, the record tested here should be browsed in sudo(), otherwise the orm could filter some set taxes you shouldn't see, no?